File: simulators/go/scenarios/prepared_statements.go
package scenarios

import (
	"context"
	"fmt"
	"sync"
	"sync/atomic"
	"time"
)

// PreparedStatements tests named and unnamed prepared statements, including reuse.
//
// Unnamed/parameterized queries (conn.QueryRow) work in all pool modes because they
// use the extended query protocol with an unnamed prepared statement that is re-parsed
// on every round-trip.
//
// Named prepared statements (conn.Prepare) create server-side state scoped to a specific
// backend. In transaction pool mode, the PREPARE, all subsequent EXECUTEs, and the final
// DEALLOCATE must run within a single BEGIN...COMMIT so that the proxy keeps the same
// backend pinned for the entire sequence. Outside a transaction the backend is released
// after each ReadyForQuery, and the next statement would land on a different backend that
// has no knowledge of the prepared statement.
func PreparedStatements(ctx context.Context, cfg *Config) (*Result, error) {
	var (
		ops       atomic.Int64
		errCount  atomic.Int64
		mu        sync.Mutex
		latencies []time.Duration
	)
	errs := NewErrorSampler(10)

	start := time.Now()
	var wg sync.WaitGroup

	for i := 0; i < cfg.Concurrency; i++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()

			conn, err := ConnectClient(ctx, cfg)
			if err != nil {
				ops.Add(1)
				errCount.Add(1)
				errs.Record(err)
				return
			}
			defer conn.Close(context.Background())

			for outerIter := 0; cfg.Continue(ctx, outerIter, 1); outerIter++ {
				// --- Unnamed prepared statements (work in all pool modes) ---
				for iter := 0; iter < 10; iter++ {
					t := time.Now()
					var count int
					err := conn.QueryRow(ctx, "SELECT count(*) FROM users WHERE balance > $1",
						float64(iter*100)).Scan(&count)
					d := time.Since(t)
					mu.Lock()
					latencies = append(latencies, d)
					mu.Unlock()
					ops.Add(1)
					if err != nil {
						if ctx.Err() != nil || conn.IsClosed() {
							return
						}
						errCount.Add(1)
						errs.Record(err)
					}
				}

				// --- Named prepared statements ---
				// Wrapped in a transaction so the proxy pins the backend for the full
				// PREPARE → EXECUTE × 5 → DEALLOCATE sequence (required in transaction mode).
				// outerIter is included in stmtName to avoid conflicts across loop iterations.
				if _, err := conn.Exec(ctx, "BEGIN"); err != nil {
					if ctx.Err() != nil || conn.IsClosed() {
						return
					}
					errCount.Add(1)
					errs.Record(err)
					ops.Add(1)
					return
				}
				rollback := func() { _, _ = conn.Exec(context.Background(), "ROLLBACK") }

				stmtName := fmt.Sprintf("stmt_worker_%d_%d", workerID, outerIter)
				t := time.Now()
				_, err = conn.Prepare(ctx, stmtName, "SELECT id, name, balance FROM users WHERE id = $1")
				d := time.Since(t)
				mu.Lock()
				latencies = append(latencies, d)
				mu.Unlock()
				ops.Add(1)
				if err != nil {
					if ctx.Err() != nil || conn.IsClosed() {
						rollback()
						return
					}
					errCount.Add(1)
					errs.Record(err)
					rollback()
					return
				}

				// Reuse the named statement many times — all on the same pinned backend.
				// Larger reuse count shifts the measurement toward EXECUTE throughput (the
				// actual purpose of prepared statements), making it less sensitive to the
				// one-time PREPARE cost and to backend plan-cache state transitions that
				// otherwise produced bimodal trial results.
				const reuseCount = 20
				for iter := 0; iter < reuseCount; iter++ {
					t = time.Now()
					rows, err := conn.Query(ctx, stmtName, workerID*reuseCount+iter+1)
					d = time.Since(t)
					mu.Lock()
					latencies = append(latencies, d)
					mu.Unlock()
					ops.Add(1)
					if err != nil {
						if ctx.Err() != nil || conn.IsClosed() {
							rollback()
							return
						}
						errCount.Add(1)
						errs.Record(err)
						continue
					}
					rows.Close()
				}

				_, err = conn.Exec(ctx, "DEALLOCATE "+stmtName)
				ops.Add(1)
				if err != nil {
					if ctx.Err() != nil || conn.IsClosed() {
						return
					}
					errCount.Add(1)
					errs.Record(err)
				}

				if _, err := conn.Exec(ctx, "COMMIT"); err != nil {
					if ctx.Err() != nil || conn.IsClosed() {
						return
					}
					errCount.Add(1)
					errs.Record(err)
				}
			}
		}(i)
	}

	wg.Wait()
	duration := time.Since(start)

	return &Result{
		TotalOps:  int(ops.Load()),
		Errors:    int(errCount.Load()),
		Duration:  duration,
		Latencies: latencies,
		Notes:     errs.Notes(),
	}, nil
}

File: simulators/typescript/src/scenarios/preparedStatements.ts
import { Pool } from "pg";
import { Config, RawResult } from "../types";

export async function preparedStatements(config: Config): Promise<RawResult> {
  const pool = new Pool({ host: config.host, port: config.port, user: config.user, password: config.password, database: config.database, max: config.concurrency });
  pool.on("error", () => {}); // suppress unhandled pool-level errors
  let ops = 0, errors = 0;
  const latencies: number[] = [];

  const start = Date.now();

  // Phase 1: unnamed parameterized queries — safe in all pool modes.
  await Promise.all(Array.from({ length: config.concurrency }, (_, i) => (async () => {
    for (let iter = 0; iter < 10; iter++) {
      const t = performance.now();
      try { await pool.query("SELECT count(*) FROM users WHERE balance > $1", [iter * 100]); }
      catch { errors++; }
      latencies.push(performance.now() - t); ops++;
    }
  })()));

  // Phase 2: named prepared statements wrapped in BEGIN/COMMIT to pin the backend.
  // In transaction mode each BEGIN holds a backend until COMMIT, so run serially to
  // avoid exhausting the (small) backend pool.  In session mode run concurrently.
  const namedWorker = async (i: number) => {
    const stmtName = `stmt_w${i}`;
    const client = await pool.connect();
    client.on("error", () => {});
    try {
      await client.query("BEGIN");
      let t = performance.now();
      try {
        await client.query(`PREPARE ${stmtName} AS SELECT id, name, balance FROM users WHERE id = $1`);
      } catch {
        errors++; ops++;
        try { await client.query("ROLLBACK"); } catch {}
        return;
      }
      latencies.push(performance.now() - t); ops++;

      for (let iter = 0; iter < 5; iter++) {
        t = performance.now();
        try { await client.query(`EXECUTE ${stmtName}(${i * 5 + iter + 1})`); }
        catch { errors++; }
        latencies.push(performance.now() - t); ops++;
      }

      try { await client.query(`DEALLOCATE ${stmtName}`); } catch { errors++; }
      ops++;
      await client.query("COMMIT");
    } catch {
      errors++;
      try { await client.query("ROLLBACK"); } catch {}
    } finally { client.release(); }
  };

  const indices = Array.from({ length: config.concurrency }, (_, i) => i);
  if (config.poolMode === "transaction") {
    // Serial: only one BEGIN/COMMIT transaction active at a time.
    for (const i of indices) await namedWorker(i);
  } else {
    await Promise.all(indices.map(namedWorker));
  }

  await pool.end();
  return { totalOps: ops, errors, durationMs: Date.now() - start, latencies };
}
